### DISCUSSION TITLE
VRF , where are the random numbers generated?

### DISCUSSION THREAD

**Discussion by r0bchain** [Cyfrin/foundry-full-course-cu#1928] [General]
Hi there, I am trying to understand this part of the course, and I was wondering where the random numbers are generated. I understand that many nodes in the Chainlink network need to validate these random numbers. But how and where are they generated? Is it a random node in the blockchain that generates the numbers, and then the other nodes validate these values? I am a bit confused with this part. Thanks.

**Comment by EngrPips** [comment]
This is an interesting question to ask. I had never bothered to even think about this since Chainlink is trusted and the random number can be verified, but honestly, this is intriguing. I will say you should read through their docs `https://docs.chain.link/vrf` because if there is anywhere we can find more details about their product, then it should be their docs. Read through it, and maybe they explained the entire process of requesting and generating the random number in their docs.

**Comment by drained69** [comment]
The many nodes serves as different oracles and collects data differently, hashes them but collectively come together to generate a random number or numbers. The different datas it collects usually contains hash of future events so its impossible to guess. For the sources of theses data, you could read this article (https://pyth.network/blog/secure-random-numbers-for-blockchains).

**Reply by EngrPips** [reply]
The article, please @Chinwuba22.

**Reply by drained69** [reply]
https://pyth.network/blog/secure-random-numbers-for-blockchains

**Reply by EngrPips** [reply]
Thank you so much, Friend.

**Reply by r0bchain** [reply]
> The many nodes serves as different oracles and collects data differently, hashes them but collectively come together to generate a random number or numbers. The different datas it collects usually contains hash of future events so its impossible to guess. For the sources of theses data, you could read this article (https://pyth.network/blog/secure-random-numbers-for-blockchains).

Thank you for the link. I was reading about the "**Verifiable Random Function**" and it makes sense, but I still don't understand why the service provider who generates the result 𝑌  (the number(s)) and p (the proof) need to do it **off-chain**. The other nodes will verify the result **on-chain** after the service provider sends 𝑌  and 𝑝 , correct? If this is true, does it mean that all validators need to recalculate the hashed proof 𝑝? If so, I don't understand why the provider cannot be an on-chain node like the validators?? I think there is something I am still missing. On the other hand, if the service provider is off-chain, it is decentralized, right? I still do not get the point.
Thanks

**Reply by drained69** [reply]
Doing it offchain makes sense as it will save costs as well.

**Reply by EngrPips** [reply]
Interesting

**Reply by r0bchain** [reply]
@Chinwuba22 Yeah, I understand that saving costs is an obvious reason, but are there other reasons? I am wondering if the number is generated by a "VRF Provider," does it mean it is generated by an isolated entity off-chain (a single computation unit)? How this Provider is selected? And after the number is generated, it is sent on-chain for node validation? If all of this is true, who can guarantee that the provider is honest and not manipulated? How is this ensured? Isn't this a form of centralization?, which is exactly what Chainlink and many other protocols aim to avoid?  Regarding the verification, I researched a bit and finally understood that with the value of X (the input provided by the requester/client), the result Y (the generated random number), and P (the proof), there is no need for the on-chain validators to recalculate the number again (thanks to the cryptographic) . This makes the verification process quick.

Thanks.
